[AOTInductor] Use ProxyExecutor for aten op if c-shim is missing

SherlockNoMad opened the pull request:
Summary:
As discussed in the meeting, we are inverting the policy on the use of proxy executor for aten fallbacks.
By default, aten fallback ops will use proxy executor, unless a c-shim is available.

Test Plan: CIs

Differential Revision: D51417683




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

chenyang78 commented on test/inductor/test_aot_inductor.py:
```diff
@@ -973,6 +973,7 @@ def forward(self, x):
         example_inputs = (torch.randn(8, 4, 4, device=self.device),)
         self.check_model(Model(), example_inputs)
 
+    @unittest.skipIf(IS_FBCODE, "Not runnable in fbcode")
```
Wondering why this test is not runnable in fbcode? Does the change of using ProxyExecutor as fallback by default affect this test? Thanks.

SherlockNoMad replied:
In OSS ABI mode, this test is failing due to aten.index.Tensor missing c-shim, this is why we have it as xfail in the first place. 
In fbcode ABI mode, this is failing due to extern_kernel_serializer not populated in AOTInductorModelRunner. I tried to add it, but this introduces dependency of fbcode module ‘deeplearning.aot_inductor.extern_node_thrift_serializer’ into an oss test file, which buck build failed. 

As a result, I added the same test case to `test_custom_op.py`, as you can see in the fbcode diff.

chenyang78 reviewed (approved):
Have a minor question. LGTM overall. Thanks.

SherlockNoMad committed:
[AOTInductor] Use ProxyExecutor for aten op if c-shim is missing (#113918)

Summary:

As discussed in the meeting, we are inverting the policy on the use of proxy executor for aten fallbacks.
By default, aten fallback ops will use proxy executor, unless a c-shim is available.

Added test for aten.sort and aten.index.Tensor, as they are now runnable with proxy executor.

Test Plan: CIs

Reviewed By: chenyang78

Differential Revision: D51417683
